packages/flutter/lib/src/rendering/object.dart

```diff
@@ -819,6 +834,146 @@ class PipelineOwner {
+
+  Map<RenderObject, List<_ElevationData>> _elevations;
+  Path _translatePath(Path p, Offset offset) {
+    final Matrix4 matrix = Matrix4.identity()..translate(offset.dx, offset.dy);
+    return p.transform(matrix.storage);
+  }
+  bool _printedExceededmaxElevationObjectsToCheckWarning = false;
+
+  /// Checks for elevation consistency with regard to painting order.
+  ///
+  /// This method is only meant for debug builds, and should only be called from
+  /// an `assert` from within the framework. All of the parameters must not be
+  /// null.
+  ///
+  /// The [elevation] refers to the elevation specified on the
+  /// [RenderPhysicalModel] or [RenderPhysicalShape]. The [area] parameter
+  /// specifies the [Rect] on the screen in which the physical model will be
+  /// drawn, and the [path] parameter specifies the actual [Path] to be drawn.
+  /// The path is expected to be smaller than the area.
+  ///
+  /// If the durrent paint cycle results in drawing a shape at an elevation
+  /// lower than another shape already drawn in this area, a [Path] will be
+  /// returned describing the boundary of the conflicting shapes. Callers can
+  /// use this path to highlight the area of the screen where the conflict
+  /// occurs.
+  ///
+  /// This check has worst case performance of O(Σn) time, which is better than
+  /// O(n^2) but worse than O(n log(n)) (its curve is similar to O(n^1.8)). If
+  /// the check degrades performance unacceptably for debug usage, consider
+  /// setting [maxElevationObjectsToCheck] to a lower number, or to 0 to disable
+  /// it completely. The number can still be set higher in unit tests, where
+  /// performance is less critical.
+  Path debugCheckElevationData(
+    double elevation,
+    Rect area,
+    Path path,
+    RenderObject object,
+    RenderObject lastCommonAncestor,
+  ) {
+    assert(elevation != null);
+    assert(area != null);
+    assert(path != null);
+    assert(object != null);
+    _elevations[lastCommonAncestor] ??= <_ElevationData>[];
+    // Check in reverse order - we're more likely to fail on a closer node
+    // if we're going to fail at all. Take only _maxElevationObjectsToCheck
+    // to avoid this taking too long in the paint cycle.
+    final Iterable<_ElevationData> elevationsToCheck = _elevations[lastCommonAncestor]
+        .reversed
+        .take(_maxElevationObjectsToCheck)
+        .where((_ElevationData elevationData) {
+          return elevation < elevationData.elevation &&
+                 area.overlaps(elevationData.area);
+        });
+    for (final _ElevationData elevationData in elevationsToCheck) {
+      final Path difference = Path.combine(
+        PathOperation.union,
+        path,
+        _translatePath(elevationData.path, elevationData.area.topLeft - area.topLeft),
+      );
+      // Check if the union created a path with a single segment. If so, the
+      // paths overlap.
+      // This will not work if we're dealing with input paths that are already
```

Comment:
I wonder how common this is going to be - and how expensive it actually would be to account for it.

Reply:
I've updated the logic to support this - if it's looking good I'll do some perf tests and we can figure out if it needs to go or maybe get guarded somehow.